File: apps/ios/DropBeam/SignalingClient.swift
import Foundation

// MARK: - Wire types (mirrors packages/protocol/src/signaling.ts)

struct DeviceInfo: Codable {
    let name: String
    let kind: String
    let peerId: String
    let version: Int

    enum CodingKeys: String, CodingKey {
        case name, kind, peerId = "peer_id", version
    }
}

// Client → Server
enum ClientMessage: Encodable {
    case createRoom(DeviceInfo)
    case joinRoom(code: String, device: DeviceInfo)
    case signal(to: String, data: [String: AnyCodable])
    case leaveRoom
    case ping

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: DynKey.self)
        switch self {
        case .createRoom(let device):
            try container.encode("create-room", forKey: .type)
            try container.encode(device, forKey: .device)
        case .joinRoom(let code, let device):
            try container.encode("join-room", forKey: .type)
            try container.encode(code, forKey: .code)
            try container.encode(device, forKey: .device)
        case .signal(let to, let data):
            try container.encode("signal", forKey: .type)
            try container.encode(to, forKey: .to)
            try container.encode(data, forKey: .data)
        case .leaveRoom:
            try container.encode("leave-room", forKey: .type)
        case .ping:
            try container.encode("ping", forKey: .type)
        }
    }
}

// Server → Client
struct ServerMessage: Decodable {
    let type: String
    let roomId: String?
    let code: String?
    let peerId: String?
    let device: DeviceInfo?
    let from: String?
    let data: [String: AnyCodable]?
    let error: String?
    let message: String?

    enum CodingKeys: String, CodingKey {
        case type, roomId = "room_id", code
        case peerId = "peer_id", device, from, data, error, message
    }
}

// MARK: - Signaling Client

typealias SignalHandler = (String, [String: AnyCodable]) -> Void

final class SignalingClient: NSObject, URLSessionWebSocketDelegate {
    private let url: URL
    private var socket: URLSessionWebSocketTask?
    private var session: URLSession?
    private var handlers: [SignalHandler] = []
    private var peerHandlers: [String: SignalHandler] = [:]

    var onRoomCreated: ((String, String) -> Void)?  // (roomId, code)
    var onRoomJoined: ((String, String) -> Void)?   // (roomId, peerId)
    var onPeerJoined: ((String, DeviceInfo) -> Void)?
    var onPeerLeft: ((String) -> Void)?
    var onError: ((String) -> Void)?

    init(url: URL) {
        self.url = url
        super.init()
        session = URLSession(configuration: .default, delegate: self, delegateQueue: nil)
    }

    func connect() {
        socket = session?.webSocketTask(with: url)
        socket?.resume()
        receive()
    }

    func disconnect() {
        socket?.cancel(with: .normalClosure, reason: nil)
    }

    func createRoom(_ device: DeviceInfo) {
        send(.createRoom(device))
    }

    func joinRoom(code: String, device: DeviceInfo) {
        send(.joinRoom(code: code, device: device))
    }

    func signal(to peerId: String, data: [String: AnyCodable]) {
        send(.signal(to: peerId, data: data))
    }

    func leaveRoom() {
        send(.leaveRoom)
    }

    func on(_ handler: @escaping SignalHandler) {
        handlers.append(handler)
    }

    func onSignalFrom(_ peerId: String, handler: @escaping SignalHandler) {
        peerHandlers[peerId] = handler
    }

    // MARK: - Private

    private func send(_ msg: ClientMessage) {
        guard let data = try? JSONEncoder().encode(msg),
              let text = String(data: data, encoding: .utf8) else { return }
        socket?.send(.string(text)) { _ in }
    }

    private func receive() {
        socket?.receive { [weak self] result in
            guard let self = self else { return }
            switch result {
            case .success(.string(let text)):
                self.handle(text)
                self.receive()
            case .success(.data(let data)):
                if let text = String(data: data, encoding: .utf8) {
                    self.handle(text)
                }
                self.receive()
            case .failure:
                break
            @unknown default:
                self.receive()
            }
        }
    }

    private func handle(_ text: String) {
        guard let data = text.data(using: .utf8),
              let msg = try? JSONDecoder().decode(ServerMessage.self, from: data) else { return }

        DispatchQueue.main.async {
            switch msg.type {
            case "room-created":
                if let id = msg.roomId, let code = msg.code {
                    self.onRoomCreated?(id, code)
                }
            case "room-joined":
                if let id = msg.roomId, let pid = msg.peerId {
                    self.onRoomJoined?(id, pid)
                }
            case "peer-joined":
                if let pid = msg.peerId, let device = msg.device {
                    self.onPeerJoined?(pid, device)
                }
            case "peer-left":
                if let pid = msg.peerId {
                    self.onPeerLeft?(pid)
                }
            case "signal":
                if let from = msg.from, let signalData = msg.data {
                    self.peerHandlers[from]?(from, signalData)
                    self.handlers.forEach { $0(from, signalData) }
                }
            case "error":
                self.onError?(msg.error ?? msg.message ?? "unknown error")
            default:
                break
            }
        }
    }
}

// MARK: - URLSessionWebSocketDelegate

extension SignalingClient {
    func urlSession(_ session: URLSession, webSocketTask: URLSessionWebSocketTask,
                    didOpenWithProtocol protocol: String?) {}
    func urlSession(_ session: URLSession, webSocketTask: URLSessionWebSocketTask,
                    didCloseWith closeCode: URLSessionWebSocketTask.CloseCode, reason: Data?) {}
}

// MARK: - Helpers

private enum DynKey: CodingKey {
    case type, device, code, to, data
    var stringValue: String {
        switch self {
        case .type: return "type"
        case .device: return "device"
        case .code: return "code"
        case .to: return "to"
        case .data: return "data"
        }
    }
    init?(stringValue: String) { return nil }
    var intValue: Int? { nil }
    init?(intValue: Int) { return nil }
}

// Type-erased Codable for JSON signal data
struct AnyCodable: Codable {
    let value: Any

    init(_ value: Any) { self.value = value }

    init(from decoder: Decoder) throws {
        let c = try decoder.singleValueContainer()
        if let v = try? c.decode(String.self) { value = v }
        else if let v = try? c.decode(Bool.self) { value = v }
        else if let v = try? c.decode(Double.self) { value = v }
        else if let v = try? c.decode([String: AnyCodable].self) { value = v }
        else if let v = try? c.decode([AnyCodable].self) { value = v }
        else { value = NSNull() }
    }

    func encode(to encoder: Encoder) throws {
        var c = encoder.singleValueContainer()
        switch value {
        case let v as String: try c.encode(v)
        case let v as Bool: try c.encode(v)
        case let v as Double: try c.encode(v)
        case let v as Int: try c.encode(v)
        case let v as [String: AnyCodable]: try c.encode(v)
        case let v as [AnyCodable]: try c.encode(v)
        default: try c.encodeNil()
        }
    }
}

File: apps/android/app/src/main/kotlin/com/dropbeam/SignalingClient.kt
package com.dropbeam

import kotlinx.coroutines.*
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.*
import okhttp3.*
import org.json.JSONObject

// Wire types mirror packages/protocol/src/signaling.ts

data class DeviceInfo(
    val name: String,
    val kind: String,
    val peerId: String,
    val version: Int = 1
)

sealed class ServerEvent {
    data class RoomCreated(val roomId: String, val code: String) : ServerEvent()
    data class RoomJoined(val roomId: String, val peerId: String) : ServerEvent()
    data class PeerJoined(val peerId: String, val device: DeviceInfo) : ServerEvent()
    data class PeerLeft(val peerId: String) : ServerEvent()
    data class Signal(val from: String, val data: JSONObject) : ServerEvent()
    data class Error(val message: String) : ServerEvent()
}

class SignalingClient(
    private val url: String,
    private val scope: CoroutineScope = CoroutineScope(Dispatchers.IO + SupervisorJob())
) {
    private val client = OkHttpClient()
    private var ws: WebSocket? = null

    private val _events = MutableSharedFlow<ServerEvent>(extraBufferCapacity = 64)
    val events: SharedFlow<ServerEvent> = _events.asSharedFlow()

    fun connect() {
        val request = Request.Builder().url(url).build()
        ws = client.newWebSocket(request, object : WebSocketListener() {
            override fun onMessage(webSocket: WebSocket, text: String) {
                scope.launch { handleMessage(text) }
            }
            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                scope.launch { _events.emit(ServerEvent.Error(t.message ?: "connection error")) }
            }
        })
    }

    fun disconnect() { ws?.close(1000, null) }

    fun createRoom(device: DeviceInfo) = send(JSONObject().apply {
        put("type", "create-room")
        put("device", device.toJson())
    })

    fun joinRoom(code: String, device: DeviceInfo) = send(JSONObject().apply {
        put("type", "join-room")
        put("code", code)
        put("device", device.toJson())
    })

    fun signal(to: String, data: JSONObject) = send(JSONObject().apply {
        put("type", "signal")
        put("to", to)
        put("data", data)
    })

    fun leaveRoom() = send(JSONObject().put("type", "leave-room"))

    private fun send(json: JSONObject) { ws?.send(json.toString()) }

    private suspend fun handleMessage(text: String) {
        val json = runCatching { JSONObject(text) }.getOrNull() ?: return
        val event = when (json.optString("type")) {
            "room-created" -> ServerEvent.RoomCreated(
                json.getString("room_id"), json.getString("code")
            )
            "room-joined" -> ServerEvent.RoomJoined(
                json.getString("room_id"), json.getString("peer_id")
            )
            "peer-joined" -> ServerEvent.PeerJoined(
                json.getString("peer_id"), json.getJSONObject("device").toDevice()
            )
            "peer-left" -> ServerEvent.PeerLeft(json.getString("peer_id"))
            "signal" -> ServerEvent.Signal(
                json.getString("from"), json.getJSONObject("data")
            )
            "error" -> ServerEvent.Error(json.optString("error", json.optString("message", "unknown")))
            else -> return
        }
        _events.emit(event)
    }
}

private fun DeviceInfo.toJson() = JSONObject().apply {
    put("name", name); put("kind", kind)
    put("peer_id", peerId); put("version", version)
}

private fun JSONObject.toDevice() = DeviceInfo(
    name = getString("name"),
    kind = getString("kind"),
    peerId = getString("peer_id"),
    version = optInt("version", 1)
)
